File: frontend/apps/ios/Pantopus/Features/CeremonialMail/CeremonialMailContent.swift
//
//  CeremonialMailContent.swift
//  Pantopus
//
//  Render-only models + enums for the T3.7 Ceremonial Mail Compose
//  wizard. The four-moment flow:
//  1. Decide who and why (recipient + intent)
//  2. Verify the address
//  3. Compose the letter (stationery + ink + body + voice + seal)
//  4. Commit and send (send timing)
//

import Foundation

/// The four ordered steps in the wizard. Numbered so the chrome
/// readout reads "1 of 4 / 2 of 4 / 3 of 4 / 4 of 4".
public enum CeremonialMailStep: Int, CaseIterable, Sendable {
    case decide = 0
    case verify
    case compose
    case commit
    case success

    public var stepNumber: Int? {
        switch self {
        case .decide: 1
        case .verify: 2
        case .compose: 3
        case .commit: 4
        case .success: nil
        }
    }

    public static let progressTotal: Int = 4
}

/// Why the letter is being sent. Drives copy + recommended
/// stationery on the compose step.
public enum CeremonialMailIntent: String, CaseIterable, Sendable, Identifiable {
    case sayHello = "say_hello"
    case congratulations
    case condolences
    case businessNote = "business_note"
    case justBecause = "just_because"

    public var id: String {
        rawValue
    }

    public var title: String {
        switch self {
        case .sayHello: "Say hello"
        case .congratulations: "Congratulations"
        case .condolences: "Condolences"
        case .businessNote: "Business note"
        case .justBecause: "Just because"
        }
    }

    public var subtitle: String {
        switch self {
        case .sayHello: "A quick warm check-in."
        case .congratulations: "Celebrate something good."
        case .condolences: "Steady through a hard moment."
        case .businessNote: "Keep it professional."
        case .justBecause: "Send a thought, no occasion needed."
        }
    }

    public var icon: PantopusIcon {
        switch self {
        case .sayHello: .send
        case .congratulations: .star
        case .condolences: .heart
        case .businessNote: .briefcase
        case .justBecause: .pencil
        }
    }
}

public enum CeremonialMailStationery: String, CaseIterable, Sendable, Identifiable {
    case classicCream = "classic_cream"
    case midnightBlue = "midnight_blue"
    case linen
    case botanical

    public var id: String {
        rawValue
    }

    public var title: String {
        switch self {
        case .classicCream: "Classic cream"
        case .midnightBlue: "Midnight blue"
        case .linen: "Linen white"
        case .botanical: "Botanical"
        }
    }
}

public enum CeremonialMailInk: String, CaseIterable, Sendable, Identifiable {
    case walnut
    case navy
    case sepia
    case forest

    public var id: String {
        rawValue
    }

    public var title: String {
        switch self {
        case .walnut: "Walnut"
        case .navy: "Navy"
        case .sepia: "Sepia"
        case .forest: "Forest"
        }
    }
}

public enum CeremonialMailSeal: String, CaseIterable, Sendable, Identifiable {
    case waxRed = "wax_red"
    case waxBlue = "wax_blue"
    case waxBlack = "wax_black"
    case none

    public var id: String {
        rawValue
    }

    public var title: String {
        switch self {
        case .waxRed: "Red wax"
        case .waxBlue: "Blue wax"
        case .waxBlack: "Black wax"
        case .none: "No seal"
        }
    }
}

public enum CeremonialMailSendTiming: String, CaseIterable, Sendable, Identifiable {
    case now
    case morning
    case tomorrow

    public var id: String {
        rawValue
    }

    public var title: String {
        switch self {
        case .now: "Send now"
        case .morning: "Tomorrow morning"
        case .tomorrow: "Tomorrow evening"
        }
    }

    public var subtitle: String {
        switch self {
        case .now: "Deliver as soon as the recipient opens Pantopus."
        case .morning: "Land in their inbox at 8 a.m. local time."
        case .tomorrow: "Land in their inbox at 6 p.m. local time."
        }
    }
}

/// Voice-postscript recording state. Recording lives in the view
/// layer; the VM only owns the resolved URL.
public enum VoicePostscriptStatus: Sendable, Hashable {
    case empty
    case recording
    case captured(localUri: String)
    case uploading
    case uploaded(remoteUrl: String)
    case error(message: String)
}

/// Outbound nav event the host should react to.
public enum CeremonialMailEvent: Sendable, Equatable {
    case dismiss
    case openMail(mailId: String)
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/ceremonial_mail/CeremonialMailContent.kt
@file:Suppress("PackageNaming")

package app.pantopus.android.ui.screens.ceremonial_mail

import androidx.compose.runtime.Immutable
import app.pantopus.android.ui.theme.PantopusIcon

/** Four-step wizard order. */
enum class CeremonialMailStep(val index: Int) {
    Decide(0),
    Verify(1),
    Compose(2),
    Commit(3),
    Success(4),
    ;

    val stepNumber: Int?
        get() =
            when (this) {
                Decide -> 1
                Verify -> 2
                Compose -> 3
                Commit -> 4
                Success -> null
            }

    companion object {
        const val PROGRESS_TOTAL: Int = 4
    }
}

enum class CeremonialMailIntent(val wire: String, val title: String, val subtitle: String, val icon: PantopusIcon) {
    SayHello("say_hello", "Say hello", "A quick warm check-in.", PantopusIcon.Send),
    Congratulations("congratulations", "Congratulations", "Celebrate something good.", PantopusIcon.Star),
    Condolences("condolences", "Condolences", "Steady through a hard moment.", PantopusIcon.Heart),
    BusinessNote("business_note", "Business note", "Keep it professional.", PantopusIcon.Briefcase),
    JustBecause("just_because", "Just because", "Send a thought, no occasion needed.", PantopusIcon.Pencil),
}

enum class CeremonialMailStationery(val wire: String, val title: String) {
    ClassicCream("classic_cream", "Classic cream"),
    MidnightBlue("midnight_blue", "Midnight blue"),
    Linen("linen", "Linen white"),
    Botanical("botanical", "Botanical"),
}

enum class CeremonialMailInk(val wire: String, val title: String) {
    Walnut("walnut", "Walnut"),
    Navy("navy", "Navy"),
    Sepia("sepia", "Sepia"),
    Forest("forest", "Forest"),
}

enum class CeremonialMailSeal(val wire: String, val title: String) {
    WaxRed("wax_red", "Red wax"),
    WaxBlue("wax_blue", "Blue wax"),
    WaxBlack("wax_black", "Black wax"),
    None("none", "No seal"),
}

enum class CeremonialMailSendTiming(val wire: String, val title: String, val subtitle: String) {
    Now("now", "Send now", "Deliver as soon as the recipient opens Pantopus."),
    Morning("morning", "Tomorrow morning", "Land in their inbox at 8 a.m. local time."),
    Tomorrow("tomorrow", "Tomorrow evening", "Land in their inbox at 6 p.m. local time."),
}

/** Voice-postscript state. The recording side lives in the view
 *  layer; the VM owns upload + resolved URL. */
sealed interface VoicePostscriptStatus {
    data object Empty : VoicePostscriptStatus

    data object Recording : VoicePostscriptStatus

    data class Captured(val localUri: String) : VoicePostscriptStatus

    data object Uploading : VoicePostscriptStatus

    data class Uploaded(val remoteUrl: String) : VoicePostscriptStatus

    data class Error(val message: String) : VoicePostscriptStatus
}

/** Outbound nav event. */
sealed interface CeremonialMailEvent {
    data object Dismiss : CeremonialMailEvent

    data class OpenMail(val mailId: String) : CeremonialMailEvent
}

@Immutable
data class CeremonialMailFormState(
    val step: CeremonialMailStep = CeremonialMailStep.Decide,
    val recipientQuery: String = "",
    val intent: CeremonialMailIntent = CeremonialMailIntent.SayHello,
    val addressConfirmed: Boolean = false,
    val returnAddressShared: Boolean = false,
    val stationery: CeremonialMailStationery = CeremonialMailStationery.ClassicCream,
    val ink: CeremonialMailInk = CeremonialMailInk.Walnut,
    val seal: CeremonialMailSeal = CeremonialMailSeal.WaxRed,
    val bodyText: String = "",
    val sendTiming: CeremonialMailSendTiming = CeremonialMailSendTiming.Now,
)
